### PR TITLE
csv-parser appears not to be thread safe when attached to a network source

### PR DESCRIPTION
**Pull request by furiel**
For issue:
https://github.com/balabit/syslog-ng/issues/1334

csv scanner object contained a few fields that are used for iterating
through columns. As multiple threads use the same scanner object,
those threads broke the state of the common iterator. This patch
separates the general scanner options and her iterator parts. The
iterator is created on the stack of the scanner parser function, thus
avoiding race condition.

Contextual data scanner also uses this scanner object, so she is also
affected by the internal api change. To maintain original mode of
operation of the contextual data scanner, the CSVScannerIterator
object is added to the CSVContextualDataRecordScanner, and the
contextual data scanner internal api is changed accordingly.

Signed-off-by: Antal Nemes <antal.nemes@balabit.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build? (admin please type: ok to test)

File: modules/csvparser/csvparser.c [hunk 1]
```diff
@@ -117,17 +117,21 @@ csv_parser_process(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_o
   CSVParser *self = (CSVParser *) s;
   LogMessage *msg = log_msg_make_writable(pmsg, path_options);
 
-  csv_scanner_input(&self->scanner, input);
-  while (csv_scanner_scan_next(&self->scanner))
+  CSVScannerIterator iterator;
+  memset(&iterator, 0, sizeof(iterator));
+
```
Comment (bazsi) [inline modules/csvparser/csvparser.c]: I think the memset() should be part of the init function.

File: lib/scanner/csv-scanner/csv-scanner.c [hunk 2]
```diff
@@ -392,43 +392,40 @@ csv_scanner_is_scan_finished(CSVScanner *self)
...
 
 void
-csv_scanner_input(CSVScanner *self, const gchar *input)
+csv_scanner_iterator_init(CSVScannerIterator *iterator, const gchar *input)
 {
-  self->src = input;
-  self->current_column = NULL;
+  iterator->src = input;
+  iterator->current_value = g_string_sized_new(128);
+  iterator->current_column = NULL;
```
Comment (bazsi) [inline lib/scanner/csv-scanner/csv-scanner.c]: I think this is a perfect example where scratch_buffers2 would avoid another allocation in the fast path.
See pr #1368

File: lib/scanner/csv-scanner/csv-scanner.h [hunk 3]
```diff
@@ -64,25 +64,28 @@ void csv_scanner_options_set_quotes(CSVScannerOptions *options, const gchar *quo
...
   GList *current_column;
   const gchar *src;
   GString *current_value;
   gchar current_quote;
+} CSVScannerIterator;
+
+typedef struct _CSVScanner
+{
+  CSVScannerOptions *options;
 } CSVScanner;
```
Comment (bazsi) [inline lib/scanner/csv-scanner/csv-scanner.h]: The naming bothers me a little, CSVScanner was already implying some form of iteration. what's the role of the two data structures:

CSVScannerIterator is used to scan through the CSV formatted line, but what is CSVScanner?

If at all possible I'd stick to the CSVScanner name for the iterator role. Why can't we get rid off CSVScanner?

**Comment by lbudai** [comment]
@bazsi: we will keep only CSVScanner :) before you entered to the room we discussed it and we concluded the same :)

File: lib/scanner/csv-scanner/csv-scanner.c (hunk 2 above)
Comment (furiel) [inline lib/scanner/csv-scanner/csv-scanner.c]: Good point thanks. Let's do this change in a separate changeset after pr is merged.
I created https://github.com/balabit/syslog-ng/pull/1421 so not to forget.

File: lib/scanner/csv-scanner/csv-scanner.c [hunk 4]
```diff
@@ -243,146 +243,146 @@ _match_string_delimiters_at_current_position(const char *input, GList *string_de
...
-csv_scanner_get_current_name(CSVScanner *self)
+csv_scanner_get_current_name(CSVScanner *scanner)
 {
-  return (const gchar *) self->current_column->data;
+  return (const gchar *) scanner->current_column->data;
 }
 
 gboolean
-csv_scanner_is_scan_finished(CSVScanner *self)
+csv_scanner_is_scan_finished(CSVScannerOptions *options, CSVScanner *scanner)
```
Comment (lbudai) [inline lib/scanner/csv-scanner/csv-scanner.c]: Could you store a pointer `options` in the CSVScanner ?

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build? (admin please type: ok to test)

**Comment by furiel** [comment]
@kira-syslogng retest please

**Comment by furiel** [comment]
@kira-syslogng retest this please

**Comment by furiel** [comment]
ok to test

File: lib/scanner/csv-scanner/csv-scanner.h [hunk 5]
```diff
@@ -80,9 +79,7 @@ gboolean csv_scanner_scan_next(CSVScanner *pstate);
 gboolean csv_scanner_is_scan_finished(CSVScanner *pstate);
 gchar *csv_scanner_dup_current_value(CSVScanner *self);
 
-void csv_scanner_input(CSVScanner *pstate, const gchar *input);
-gboolean csv_scanner_parse_input(CSVScanner *pstate);
-void csv_scanner_state_init(CSVScanner *pstate, CSVScannerOptions *options);
+void csv_scanner_init(CSVScanner *pstate, CSVScannerOptions *options, const gchar *input);
 void csv_scanner_state_clean(CSVScanner *pstate);
```
Comment (bazsi) [inline lib/scanner/csv-scanner/csv-scanner.h]: isn't state_clean the pair of csv_scanner_init()? the naming does not suggest that.

**Comment by kira-syslogng** [comment]
Build SUCCESS
